### PR TITLE
Stateless Block Proof Construction + Proof Verifier HTTP server + "cross-validation" Against Self 

### PR DESCRIPTION
**Pull request by jwasinger**
Draft/working branch.

Implements the following functionality:

* If enabled (via setting `--crossvalidation.endpoint`), Geth constructions stateless block proofs from fully-executed blocks.  These are verified against the stateless block proof verifier included in this PR.  If failures happen, an error is logged and the rlp-encoded proofs are stored to disk.
* **stateless verifier (`cmd/stateless`)** implements an HTTP server which provides a single API method to verify a stateless block proof.

Stateless proof format is RLP-encoded struct containing a block and auxiliary data necessary to fully execute the block.  This auxiliary data contains (among other things), a set of path-node key-value pairs for state and storage tries.

Most mainnet blocks currently fail to validate.  One reason is that storage values and intermediate trie nodes are not gathered for accounts that are accessed but not mutated.  Need to look at adding a method to trace/gather existing nodes in the trie without needing to commit.

### CODE REVIEW

**Comment by jwasinger** [comment]
This supersedes https://github.com/ethereum/go-ethereum/pull/28640

File: cmd/stateless/main.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+var (
+	BlockWitnessFlag = &cli.StringFlag{
+		Name:  "block-witness",
+		Usage: "foo bar",
+	}
+
+	BlockWitness1Flag = &cli.StringFlag{
+		Name:  "witness1",
+		Usage: "foo bar",
```
Comment (holiman) [inline cmd/stateless/main.go:28]: Please fix these up. For all of these: 
- Remove the ones that aren't needed / wanted
- Add proper usage-descriptions
- Use proper names

It doesn't take a lot of time to do, and it makes the work of reviewing a lot easier if there't not a bunch of foos and bars all around

File: cmd/stateless/main.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+		debug.Exit()
+		prompt.Stdin.Close() // Resets terminal mode.
+		return nil
+	}
+}
+
+func statelessCmd(ctx *cli.Context) error {
+	blockWitnessPath := ctx.String(BlockWitnessFlag.Name)
+	if blockWitnessPath == "" {
+		panic("block witness required")
```
Comment (holiman) [inline cmd/stateless/main.go:109]: Isn't exit with `1` sufficient? You can just return the error and it will be taken care of. Having a stacktrace at this point is moot, we know what the error is and what caused it

File: cmd/stateless/main.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	correct, err := statelessVerify(witness)
+	if err != nil {
+		panic(err)
+	}
+	if !correct {
+		panic("not correct")
```
Comment (holiman) [inline cmd/stateless/main.go:127]: ... same for all of these. Remember HGTTG

File: cmd/stateless/main.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+	receipts, logs, usedGas, err := processor.Process(witness.Block, db, vmConfig)
+	if err != nil {
+		return false, err
+	}
+
+	// TODO: enumerate failure cases and divide them into:
+	// * errors that are caused by invalidly-formed proof
+	// * errors that could be consensus bugs
+	_ = logs
```
Comment (holiman) [inline cmd/stateless/main.go:165]: what is this?

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -912,12 +912,13 @@ Please note that --` + MetricsHTTPFlag.Name + ` must be set to start the server.
 
 var CrossValidationEndpointFlag = &cli.StringFlag{
 	Name:  "crossvalidation.endpoint",
-	Usage: "http endpoints to cross validate blocks against",
+	Usage: "http endpoint(s) to cross validate blocks against.  Formatted as a comma-separated list of URLs",
```
Comment (holiman) [inline cmd/utils/flags.go:915]: If you turn it into a `StringSliceFlag` you get that functionality for free. It can then be input either as
` --crossvalidation.endpoint="foo,bar,gazonk"` _or_ as `--crossvalidation.endpoint="foo"  --crossvalidation.endpoint="bar"  --crossvalidation.endpoint="gazonk"`

File: tests/block_test.go [hunk 6]
```diff
@@ -61,6 +65,85 @@ func TestBlockchain(t *testing.T) {
...
+	bt.skipLoad(`.*randomStatetest94.json.*`)
+
+	// skip uncle tests for stateless
+	bt.skipLoad(`.*/UnclePopulation.json`)
+
+	bt.walk(t, blockTestDir, func(t *testing.T, name string, test *BlockTest) {
+		if runtime.GOARCH == "386" && runtime.GOOS == "windows" && rand.Int63()%2 == 0 {
+			t.Skip("test (randomly) skipped on 32-bit windows")
+		}
+		if networkPostMerge(test.json.Network) {
```
Comment (holiman) [inline tests/block_test.go:136]: Actually, you could remove that `networkPostMerge` with something like this instead: 
```golang
		config, ok := Forks[t.json.Network]
		if !ok {
			return UnsupportedForkError{t.json.Network}
		}
		isMerged := config.TerminalTotalDifficulty != nil && config.TerminalTotalDifficulty.BitLen() == 0
		if !isMerged {
```

File: light/trie.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
+// Copyright 2015 The go-ethereum Authors
```
Comment (holiman) [inline light/trie.go:1]: Why did you add back this file? 
:)

**Review by holiman — COMMENTED** [review]
Some nits here and there, and please also fix the imports, they are non-`goimports`-compliant pretty much everywhere.

**Review by jwasinger — COMMENTED** [review]
(see below.  for some reason I'm unable to delete this comment)

File: core/state/state_object.go [hunk 8]
```diff
@@ -196,9 +200,12 @@ func (s *stateObject) GetCommittedState(key common.Hash) common.Hash {
 	)
 
 	if s.db.witness != nil {
-		// when collecting witness data with snapshot enabled, always prefetch every read key to
-		// ensure that read storage slots will end up in the witness
-		s.db.prefetcher.prefetch(s.addrHash, s.data.Root, s.address, [][]byte{key[:]})
+		if !s.createdBlock && s.origin.Root != types.EmptyRootHash {
```
Comment (jwasinger) [inline core/state/state_object.go:203]: @rjl493456442 @holiman previous to this commit, selfdestructing+recreating an account would cause the prefetcher to load the storage keys using the current root (always empty post byzantium).  Nothing would be loaded.

The error wasn't testable, and this commit is mostly cosmetic to make the logic around invocation of the prefetcher for storage reads more correct.

Comment (holiman) [reply core/state/state_object.go:203]: I don't see how it can be correct to add an unjournalled flag to stateobjects though. I mean, if we add some more 'stateful thing' on stateobjects, then surely that stateful-ness must be journalled?

Comment (holiman) [reply core/state/state_object.go:203]: How do you mean it's "mostly cosmetic" ? If it's "mostly cosmetic", it seems like an extremely high price to pay for a cosmetic change (because another created-flag on the stateobject makes a complicated thing even more complicated)

Comment (jwasinger) [reply core/state/state_object.go:203]: > I don't see how it can be correct to add an unjournalled flag to stateobjects though. I mean, if we add some more 'stateful thing' on stateobjects, then surely that stateful-ness must be journalled?

oof. yeah.

> How do you mean it's "mostly cosmetic" ?

The previous code worked without error because a state object's root is only modified (set to `0x00..0`)during block processing if it is deleted and resurrected.  So it just means that the prefetcher would stop retrieving newly read storage keys for the account, which is the correct behavior.

The changes in this commit aim to make the code more intuitive.  But I guess adding overhead and complexity to statedb isn't worth it.  I'd opt to keep the old logic and add a comment blurb about why it is correct.

File: eth/ethconfig/config.go [hunk 9]
```diff
@@ -171,9 +173,6 @@ func CreateConsensusEngine(config *params.ChainConfig, db ethdb.Database) (conse
 	}
 	// If defaulting to proof-of-work, enforce an already merged network since
 	// we cannot run PoW algorithms anymore, so we cannot even follow a chain
-	// not coordinated by a beacon node.
-	if !config.TerminalTotalDifficultyPassed {
-		return nil, errors.New("ethash is only supported as a historical component of already merged networks")
-	}
+	// not coordinated by a beacon node.j
```
Comment (karalabe) [inline eth/ethconfig/config.go:176]: Feels like this was an accidental delete?

File: trie/trie.go [hunk 10]
```diff
@@ -106,7 +106,7 @@ func NewEmpty(db *Database) *Trie {
 }
 
 // MustNodeIterator is a wrapper of NodeIterator and will omit any encountered
-// error but just print out an error message.
+// error but just printg out an error message.
```
Comment (karalabe) [inline trie/trie.go:109]: ?

File: core/state/trie_prefetcher.go [hunk 11]
```diff
@@ -251,42 +242,30 @@ func newSubfetcher(db Database, state common.Hash, owner common.Hash, root commo
...
-	select {
-	case <-sf.stop:
-	default:
-		close(sf.stop)
+func (sf *subfetcher) wait() {
+	// Signal termination by nil tasks
+	sf.lock.Lock()
+	if sf.closing {
+		sf.lock.Unlock()
+		return // already exiting
```
Comment (karalabe) [inline core/state/trie_prefetcher.go:262]: This will not `wait`, rather return early, even if shutdown is still being processed (and someone else requested wait already).

File: core/state/statedb.go [hunk 12]
```diff
@@ -866,27 +905,58 @@ func (s *StateDB) Finalise(deleteEmptyObjects bool) {
 	s.clearJournalAndRefund()
 }
 
+func (s *StateDB) collectReadStorageAccessLists() {
```
Comment (karalabe) [inline core/state/statedb.go:908]: This will only work if snapshots are enabled, no? Otherwise the readPrefetcher is a noop and reads are accumulated through the actual state object trie. So we'd need to pull the access list from that?

Comment (karalabe) [reply core/state/statedb.go:908]: Annoyingly, it could even happen in theory that some items are looked up from the state snapshot and some from the trie (if snapshot lookup fails), so we always need to merge the two.

File: core/state/statedb.go [hunk 13]
```diff
@@ -604,9 +634,15 @@ func (s *StateDB) getDeletedStateObject(addr common.Address) *stateObject {
...
 	}
+
 	// Insert into the live set
 	obj := newObject(s, addr, data)
 	s.setStateObject(obj)
+	if s.witness != nil && s.snap != nil {
+		// when building witness with snap enabled, prefetch all read accounts to later be collected and
+		// included in the witness
+		s.readPrefetcher.prefetch(common.Hash{}, s.originalRoot, common.Address{}, [][]byte{addr[:]})
+	}
```
Comment (karalabe) [inline core/state/statedb.go:645]: This won't capture the proofs for empty account reads, those get short circut above.

File: core/state/trie_prefetcher.go [hunk 14]
```diff
@@ -215,13 +209,12 @@ type subfetcher struct {
...
-	lock  sync.Mutex // Lock protecting the task queue
+	tasks   [][]byte   // Items queued up for retrieval
+	lock    sync.Mutex // Lock protecting the task queue
+	closing bool       // set to true if the subfetcher is closing
 
-	wake chan struct{}  // Wake channel if a new task is scheduled
-	stop chan struct{}  // Channel to interrupt processing
-	term chan struct{}  // Channel to signal interruption
-	copy chan chan Trie // Channel to request a copy of the current trie
+	wake chan bool     // Wake channel if a new task is scheduled, true if the subfetcher should continue running when there are no pending tasks
```
[the remark this answers is not in the document]
Comment (jwasinger) [reply core/state/trie_prefetcher.go:216]: I can't recall the original reasoning, and the code got changed in the meantime.  Looks like it would be better to move closing/wake into two separate channels.

File: core/state/statedb.go (hunk 13 above)
Comment (jwasinger) [reply core/state/statedb.go:645]: I can create a state test to try and trigger this.

File: core/state/trie_prefetcher.go (hunk 11 above)
Comment (jwasinger) [reply core/state/trie_prefetcher.go:262]: It says in the comment description that this function is not thread safe.

**Comment by lispc** [comment]
Is this idea discarded? Or moved elsewhere?
